src/librbd/mirror/snapshot/UnlinkPeerRequest.cc

```diff
@@ -61,42 +61,55 @@ void UnlinkPeerRequest<I>::unlink_peer() {
   CephContext *cct = m_image_ctx->cct;
 
   m_image_ctx->image_lock.lock_shared();
+  int r = -ENOENT;
+  cls::rbd::MirrorSnapshotNamespace* mirror_ns = nullptr;
+  bool newer_mirror_snapshots = false;
+  for (auto snap_it = m_image_ctx->snap_info.find(m_snap_id);
+       snap_it != m_image_ctx->snap_info.end(); ++snap_it) {
+    if (snap_it->first == m_snap_id) {
+      r = 0;
+      mirror_ns = boost::get<cls::rbd::MirrorSnapshotNamespace>(
+        &snap_it->second.snap_namespace);
+    } else if (boost::get<cls::rbd::MirrorSnapshotNamespace>(
+                 &snap_it->second.snap_namespace) != nullptr) {
+      newer_mirror_snapshots = true;
+      break;
+    }
+  }
 
-  auto snap_info = m_image_ctx->get_snap_info(m_snap_id);
-  if (!snap_info) {
+  if (r == -ENOENT) {
     m_image_ctx->image_lock.unlock_shared();
-    finish(-ENOENT);
+    finish(r);
     return;
   }
 
-  auto info = boost::get<cls::rbd::MirrorSnapshotNamespace>(
-    &snap_info->snap_namespace);
-  if (info == nullptr) {
-    lderr(cct) << "not mirror primary snapshot (snap_id=" << m_snap_id << ")"
-               << dendl;
+  if (mirror_ns == nullptr) {
+    lderr(cct) << "not mirror snapshot (snap_id=" << m_snap_id << ")" << dendl;
     m_image_ctx->image_lock.unlock_shared();
     finish(-EINVAL);
     return;
   }
 
-  if (info->mirror_peer_uuids.count(m_mirror_peer_uuid) == 0 ||
-      info->mirror_peer_uuids.size() == 1U) {
+  // if there is or will be no more peers in the mirror snapshot and we have
+  // a more recent mirror snapshot, remove the older one
```

Comment:
It would be nice to have this indicated somehow on the header file's diagram. Though it might be too verbose for the diagram so I don't mind much.

Reply:
Updated diagram